gammapy/modeling/fit.py

```diff
@@ -241,7 +241,8 @@ def covariance(self, datasets):
         result : `CovarianceResult`
             Results
         """
-        datasets, parameters = self._parse_datasets(datasets=datasets)
+        datasets, _ = self._parse_datasets(datasets=datasets)
+        parameters = datasets.models.parameters
```

Comment:
I'd be surprised if this worked. Directly accessing `datasets.models.parameters` will keep the shared parameter duplicated in the parameter list (as opposed to https://github.com/gammapy/gammapy/blob/master/gammapy/datasets/core.py#L147). This causes the optimizer to see the shared parameter as independent parameters, which makes them correlated 100%. In any case we should start with a unit test showing the parameter sharing works with the fit here: https://github.com/gammapy/gammapy/blob/master/gammapy/modeling/tests/test_fit.py

Then we can reproduce the issue reported in Slack and find the proper fix.

Reply:
Using `parameters `or `unique_parameters` in `compute` makes no difference (entries for duplicated parameters will be set to zero anyway), but `from_factor_matrix` need `parameters ` as input in order to expand correctly the `factor_matrix`